src/psbt.cpp

```diff
@@ -209,23 +228,77 @@ bool PSBTOutput::IsNull() const
+    if (!value && value_commitment.IsNull()) return false;
+    if (asset.IsNull() & asset_commitment.IsNull()) return false;
+
+    // Cannot have explicit value and value commitment
+    if (value && !value_commitment.IsNull()) return false;
+    // Cannot have explicit asset and asset commitment
+    if (!asset.IsNull() && !asset_commitment.IsNull()) return false;
+
+    // If this output needs to be blinded, make sure it is either fully blinded or not blinded
+    if (IsBlinded() && (IsPartiallyBlinded() && !IsFullyBlinded())) return false;
+
+    return true;
+}
+
+bool PSBTOutput::Merge(const PSBTOutput& output)
 {
     hd_keypaths.insert(output.hd_keypaths.begin(), output.hd_keypaths.end());
     unknown.insert(output.unknown.begin(), output.unknown.end());
 
     if (redeem_script.empty() && !output.redeem_script.empty()) redeem_script = output.redeem_script;
     if (witness_script.empty() && !output.witness_script.empty()) witness_script = output.witness_script;
 
+    // Check if blinding data matches if both have them
+    if (blinding_pubkey.IsValid() && output.blinding_pubkey.IsValid() && (blinding_pubkey != output.blinding_pubkey)) return false;
+    if (value && output.value && (value != output.value)) return false;
+    if (!value_commitment.IsNull() && !output.value_commitment.IsNull() && (value_commitment != output.value_commitment)) return false;
+    if (!asset_commitment.IsNull() && !output.asset_commitment.IsNull() && (asset_commitment != output.asset_commitment)) return false;
+    if (!range_proof.empty() && !output.range_proof.empty() && (range_proof != output.range_proof)) return false;
+    if (!surjection_proof.empty() && !output.surjection_proof.empty() && (surjection_proof != output.surjection_proof)) return false;
+
+    // If the current one has missing data, then combine from the incoming.
     if (!blinding_pubkey.IsValid() && output.blinding_pubkey.IsValid()) blinding_pubkey = output.blinding_pubkey;
+    if (!value && !output.value) value = output.value;
     if (value_commitment.IsNull() && !output.value_commitment.IsNull()) value_commitment = output.value_commitment;
-    if (value_blinding_factor.IsNull() && !output.value_blinding_factor.IsNull()) value_blinding_factor = output.value_blinding_factor;
     if (asset_commitment.IsNull() && !output.asset_commitment.IsNull()) asset_commitment = output.asset_commitment;
-    if (asset_blinding_factor.IsNull() && !output.asset_blinding_factor.IsNull()) asset_blinding_factor = output.asset_blinding_factor;
-    if (nonce_commitment.IsNull() && !output.nonce_commitment.IsNull()) nonce_commitment = output.nonce_commitment;
     if (range_proof.empty() && !output.range_proof.empty()) range_proof = output.range_proof;
     if (surjection_proof.empty() && !output.surjection_proof.empty()) surjection_proof = output.surjection_proof;
+
+    return true;
+}
+
+bool PSBTOutput::IsBlinded() const
+{
+    return blinding_pubkey.IsValid();
```

Comment:
Nor `IsPartiallyBlinded` neither `IsFullyBlinded` check `blinding_pubkey.IsValid()`. That means that there is a logical contradiction that an output can be 'FullyBlinded', but not 'Blinded'. This can create confusion that can result in mistakes. Maybe better wording can help, like 'Blindable' instead of 'Blinded' ? Or `IsPartiallyBlinded`/`IsFullyBlinded` should imply `IsBlinded`

Reply:
Enforced it.